File: solutions/cpp/longest-subarray-with-distinct-entries.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <unordered_map>
#include <unordered_set>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::generate_n;
using std::max;
using std::random_device;
using std::string;
using std::stoi;
using std::uniform_int_distribution;
using std::unordered_map;
using std::unordered_set;
using std::vector;

// @include
int LongestSubarrayWithDistinctEntries(const vector<int>& A) {
  // Records the last occurrence of each entry.
  unordered_map<int, size_t> last_occurrence;
  size_t starting_idx = 0, ans = 0;
  for (size_t i = 0; i < A.size(); ++i) {
    auto result = last_occurrence.emplace(A[i], i);
    if (!result.second) {  // A[i] appeared before. Check its validity.
      if (result.first->second >= starting_idx) {
        ans = max(ans, i - starting_idx);
        starting_idx = result.first->second + 1;
      }
      result.first->second = i;
    }
  }
  ans = max(ans, A.size() - starting_idx);
  return ans;
}
// @exclude

// O(n^2) checking solution.
int CheckAns(const vector<int>& A) {
  size_t len = 0;
  for (size_t i = 0; i < A.size(); ++i) {
    unordered_set<int> table;
    size_t j;
    for (j = i; A.size() - i > len && j < A.size(); ++j) {
      if (!table.emplace(A[j]).second) {
        break;
      }
    }
    len = max(len, j - i);
  }
  return len;
}

int main(int argc, char** argv) {
  default_random_engine gen((random_device())());
  uniform_int_distribution<size_t> dis(0, 1000);
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    n = dis(gen);
  }
  cout << "n = " << n << endl;
  for (int times = 0; times < 1000; ++times) {
    vector<int> A;
    generate_n(back_inserter(A), n, [&] { return dis(gen); });
    /*
    for (int a : A) {
      cout << a << " ";
    }
    cout << endl;
    */
    int ans = LongestSubarrayWithDistinctEntries(A);
    int golden_ans = CheckAns(A);
    cout << ans << " " << golden_ans << endl;
    assert(ans == golden_ans);
  }
  return 0;
}

File: solutions/python/longest_subarray_with_distinct_entries.py
import sys
import random


# @include
def longest_subarray_with_distinct_entries(A):
    # Records the most recent occurrences of each entry.
    most_recent_occurrence = {}
    longest_dup_free_subarray_start_idx = result = 0
    for i, a in enumerate(A):
        # Defer updating dup_idx until we see a duplicate.
        if a in most_recent_occurrence:
            dup_idx = most_recent_occurrence[a]
            # A[i] appeared before. Did it appear in the longest current
            # subarray?
            if dup_idx >= longest_dup_free_subarray_start_idx:
                result = max(result, i - longest_dup_free_subarray_start_idx)
                longest_dup_free_subarray_start_idx = dup_idx + 1
        most_recent_occurrence[a] = i
    return max(result, len(A) - longest_dup_free_subarray_start_idx)
# @exclude


# O(n^2) checking solution.
def check_ans(A):
    length = 0
    for i in range(len(A)):
        table = set()
        j = i
        while len(A) - i > length and j < len(A):
            if A[j] in table:
                break
            table.add(A[j])
            j += 1
        length = max(length, j - i)
    return length


def simple_test():
    assert 1 == longest_subarray_with_distinct_entries([1, 1, 1])
    assert 2 == longest_subarray_with_distinct_entries([1, 2, 1])
    assert 3 == longest_subarray_with_distinct_entries([1, 2, 1, 3, 1, 2, 1])
    assert 2 == longest_subarray_with_distinct_entries(
        [1, 2, 2, 3, 3, 1, 1, 2, 1])


def main():
    simple_test()
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(0, 1000)
    print('n =', n)
    for _ in range(1000):
        A = [random.randint(0, 1000) for _ in range(n)]
        ans = longest_subarray_with_distinct_entries(A)
        golden_ans = check_ans(A)
        print(ans, golden_ans)
        assert ans == golden_ans


if __name__ == '__main__':
    main()
